src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +20,66 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return BATTERY_AMP_HOUR_CAPACITY-AmpHours;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (Current < 0)
+    {
+        return false;
+    }
+    else
+    {
+        return true;
+    }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    double timeRemaining = (BATTERY_AMP_HOUR_CAPACITY-AmpHours)/inCurrent;
+    timeRemaining = qAbs(timeRemaining)/2.77778e-7;
+    return QTime(0, 0, 0, 0).addMSecs(timeRemaining);
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
+
+    QTime currentTime = batteryData.time;
+    changeTime = abs(initialTime.msecsTo(currentTime));
+    changeTime = changeTime * 2.77778e-7;
```

Comment:
For string and numeric constants, it is a good programming practice to extract them out into named constants. In c++ one way of doing it is to use anonymous namespaces near the top of the file. The syntax would be like:

```c++
namespace
{
   const QString SOME_CONVERSION_FACTOR = 2.77778e-7;
}
```

The benefit for doing it this way is that if the format ever changes then you don't need to change it in multiple places.